tree-walk.go

```diff
@@ -37,43 +37,104 @@ type treeWalkResult struct {
+		}
+		// For each rune in the entry string, if any rune is less than '/'
+		// then the next entry should not contain '/' at the same same position.
+		for j := 0; j < len(entry); j++ {
+			if entry[j] < '/' {
+				if len(entries[i+1]) > j {
+					if entries[i+1][j] == '/' {
+						return false
+					}
+				}
+			}
+		}
+	}
+	return true
+}
+
 // "listDir" function of type listDirFunc returned by listDirFactory() - explained below.
-type listDirFunc func(bucket, prefixDir, prefixEntry string) (entries []string, err error)
+type listDirFunc func(bucket, prefixDir, prefixEntry string) (entries []string, delayIsLeaf bool, err error)
+
+// A function isLeaf of type isLeafFunc is used to detect if an entry is a leaf entry. There are four scenarios
+// where isLeaf should behave differently:
+// 1. FS backend object listing - isLeaf is true if the entry has a trailing "/"
+// 2. FS backend multipart listing - isLeaf is true if the entry is a directory and contains uploads.json
+// 3. XL backend object listing - isLeaf is true if the entry is a directory and contains xl.json
+// 4. XL backend multipart listing - isLeaf is true if the entry is a directory and contains uploads.json
+type isLeafFunc func(string, string) bool
 
 // Returns function "listDir" of the type listDirFunc.
 // isLeaf - is used by listDir function to check if an entry is a leaf or non-leaf entry.
 // disks - used for doing disk.ListDir(). FS passes single disk argument, XL passes a list of disks.
-func listDirFactory(isLeaf func(string, string) bool, disks ...StorageAPI) listDirFunc {
+func listDirFactory(isLeaf isLeafFunc, disks ...StorageAPI) listDirFunc {
 	// listDir - lists all the entries at a given prefix and given entry in the prefix.
-	// isLeaf is used to detect if an entry is a leaf entry. There are four scenarios where isLeaf
-	// should behave differently:
-	// 1. FS backend object listing - isLeaf is true if the entry has a trailing "/"
-	// 2. FS backend multipart listing - isLeaf is true if the entry is a directory and contains uploads.json
-	// 3. XL backend object listing - isLeaf is true if the entry is a directory and contains xl.json
-	// 4. XL backend multipart listing - isLeaf is true if the entry is a directory and contains uploads.json
-	listDir := func(bucket, prefixDir, prefixEntry string) (entries []string, err error) {
+	listDir := func(bucket, prefixDir, prefixEntry string) (entries []string, delayIsLeaf bool, err error) {
 		for _, disk := range disks {
 			if disk == nil {
 				continue
 			}
 			entries, err = disk.ListDir(bucket, prefixDir)
 			if err == nil {
-				// Skip the entries which do not match the prefixEntry.
-				for i, entry := range entries {
-					if !strings.HasPrefix(entry, prefixEntry) {
-						entries[i] = ""
-						continue
+				// Sort and filter out the entries that does not have prefix prefixEntry.
+				sort.Strings(entries)
+				start := 0
+				end := len(entries)
+				for {
+					if start == end {
+						break
 					}
```

Comment:
This code block has become bigger, we should move this a function which basically just skips all entries which do not have the prefix.